MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs

```diff
@@ -0,0 +1,136 @@
+#pragma warning disable CS1591
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Diacritics.Extensions;
+using MediaBrowser.Controller.Entities.Audio;
+using MediaBrowser.Controller.Providers;
+using MediaBrowser.Model.Entities;
+using MediaBrowser.Model.Providers;
+using MediaBrowser.Providers.Plugins.MusicBrainz;
+using MetaBrainz.MusicBrainz;
+using MetaBrainz.MusicBrainz.Interfaces.Entities;
+using Microsoft.Extensions.Logging;
+
+namespace MediaBrowser.Providers.Music
+{
+    public class MusicBrainzArtistProvider : IRemoteMetadataProvider<MusicArtist, ArtistInfo>
+    {
+        private Query _musicBrainzQuery;
+
+        public MusicBrainzArtistProvider()
+        {
+            _musicBrainzQuery = new Query();
+
+            Current = this;
+        }
+
+        internal static MusicBrainzArtistProvider Current { get; private set; }
+
+        /// <inheritdoc />
+        public string Name => "MusicBrainz";
+
+        /// <inheritdoc />
+        public async Task<IEnumerable<RemoteSearchResult>> GetSearchResults(ArtistInfo searchInfo, CancellationToken cancellationToken)
+        {
+            var artistId = searchInfo.GetMusicBrainzArtistId();
+
+            if (!string.IsNullOrWhiteSpace(artistId))
+            {
+                var artistResult = await _musicBrainzQuery.LookupArtistAsync(new Guid(artistId)).ConfigureAwait(false);
+                return GetResultsFromResponse(new[] { artistResult });
+            }
+            else
+            {
+                // They seem to throw bad request failures on any term with a slash
+                var nameToSearch = searchInfo.Name.Replace('/', ' ');
+
+                var artistSearchResults = await _musicBrainzQuery.FindArtistsAsync($"\"{searchInfo.Name}\"");
+                if (artistSearchResults.Results.Count > 0)
+                {
+                    return GetResultsFromResponse(artistSearchResults.Results.Select(result => result.Item).ToList());
+                }
+
+                if (searchInfo.Name.HasDiacritics())
+                {
+                    // Try again using the search with an accented characters query
+                    var artistAccentsSearchResults = await _musicBrainzQuery.FindArtistsAsync($"artistaccent:\"{searchInfo.Name}\"");
+                    if (artistAccentsSearchResults.Results.Count > 0)
+                    {
+                        return GetResultsFromResponse(artistAccentsSearchResults.Results.Select(result => result.Item).ToList());
```

Comment:
I'm generally not a fan of the use of `GetResultsFromResponse`. You should have a method that takes a single `IArtist` and returns a new `RemoteSearchResult` and `GetResultsFromResponse` should yield return this method.

In the current design there's some wasted allocations. You're calling ToList (list is allocated) and then calling Select (enumerator is created and then further up the call stack ToList is called again), which is incorrect use of IEnumerable. And for single values you're allocating a new array, which is a waste.